### PR TITLE
[llvm][CodeGen] Added a check in CodeGenPrepare::optimizeSwitchType 

### PR DESCRIPTION
**Pull request by karouzakisp**
When we have two or more users in the Switch Condition and we have more SExt Users than ZExt Users or more signed cmp Instruction Users than unsigned cmp Instruction Users, we need to check to not insert a ZExt into the switch condition Type and rather insert a SExt. This way we can eliminate one Instruction.

### CODE REVIEW

**Comment by arsenm** [comment]
Can you rephrase the description to say what the check is, and why?

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 1]
```diff
@@ -7397,7 +7397,19 @@ bool CodeGenPrepare::optimizeSwitchType(SwitchInst *SI) {
     if (Arg->hasZExtAttr())
       ExtType = Instruction::ZExt;
   }
-
+  // If Cond has 2 users and one of them is a sext don't use a zext.
```
Comment (preames) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:7400]: We should probably be looking at the other user and using it's extend kind (sign or zero).

we probably do want the argument extend kind check to overrule, so maybe put this just above?

File: llvm/test/CodeGen/RISCV/switch-width.ll [hunk 2]
```diff
@@ -194,25 +194,108 @@ return:
   %retval = phi i32 [ -1, %sw.default ], [ 0, %sw.bb0 ], [ 1, %sw.bb1 ]
   ret i32 %retval
 }
+define i32 @sext_i32(i16 %a)  {
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/switch-width.ll:197]: You should add test variants here with zeroext and signext on the arguments to make sure we test the interaction.

**Review by preames — CHANGES_REQUESTED** [review]

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 3]
```diff
@@ -7387,17 +7387,38 @@ bool CodeGenPrepare::optimizeSwitchType(SwitchInst *SI) {
...
+    }
+  }
+  if (SExtUsers > ZExtUsers) {
+    ExtType = Instruction::SExt;
+  } else {
+    ExtType = Instruction::ZExt;
   }
 
+  // Some targets prefer SExt over ZExt.
+  if (TLI->isSExtCheaperThanZExt(OldVT, RegType))
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:7419]: What was the point of doing this user scan if you just throw away the result?

Comment (karouzakisp) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:7419]: Because in some cases we have more ZExtUsers but SExt is cheaper so if create a ZExt we add more instructions. 
I checked spec2017 for RISC-V with the precedence with this and It was doing worse on 500.perlbench_r if we check SExtIsCheaperThanZExt before the user scan. I can share the results later

Comment (karouzakisp) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:7419]: This is the text sizes with SExtIsCheaperThanZExt take no precedence over the user scan 
SextIsCheaperThanZextisBeforeUserScan
And this is the text sizes with SExtIsCheaperThanZExt taking precedence than user scan.
SextIsCheaperThanZextisAfterUserScan

Comment (arsenm) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:7419]: I don't mean the end result, I mean the computation you perform in this code is discarded based on isSExtCheaperThanZExt

**Review by arsenm — COMMENTED** [review]
Title should be more specific about the purpose of the check, rather than state that a check was added

File: llvm/lib/CodeGen/CodeGenPrepare.cpp (hunk 3 above)
Comment (karouzakisp) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:7419]: Yes isSExtCheaperThanZExt needs to take precedence if it doesn't agree with the computation that I perform otherwise the code size increases as I showed above.
Maybe the gain is very small compared to the cost because we need a User Scan.

**Review by arsenm — CHANGES_REQUESTED** [review]
Title needs improvement

File: llvm/lib/CodeGen/CodeGenPrepare.cpp (hunk 3 above)
Comment (michaelmaitland) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:7419]: My reading of @arsenm's comment is that maybe it would be better to avoid the user scan if we dont need it as such:
```
if (TLI->isSExtCheaperThanZExt(OldVT, RegType))
  ExtType = Instruction::SExt;
else {
 // User scan above
}

Comment (topperc) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:7419]: If one of the users is a ZExt, then ZExt is preferable to SExt even if SExt is cheaper.
